printer/location_tracker_formatter.ml

```diff
@@ -0,0 +1,226 @@
+(*                                                                        *)
+(*   Copyright 2018--2019 Jane Street Group LLC                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+let is_activated = ref false
+
+let activate_tracking () = is_activated := true
+let deactivate_tracking () = is_activated := false
+    let is_tracking () = !is_activated
+
+module Location_in_file = struct
+  type t = {
+    line : int;
+    char : int;
+    bol : int;
+  }
+
+  let print ppf { line; char; bol } = Format.fprintf ppf "{line=%d; char=%d; bol=%d}" line char bol
+
+  let create ~line ~char ~bol =
+    { line; char; bol }
+
+  let line { line; _ } = line
+  let char { char; _ } = char
+  let bol { bol; _ } = bol
+end
+
+module Range = struct
+  type t = {
+    file : string;
+    start : Location_in_file.t;
+    end_ : Location_in_file.t;
+  }
+
+  let print ppf {file; start; end_} =
+    Format.fprintf ppf "{file=%s; start=%a; end=%a}" file Location_in_file.print start Location_in_file.print end_
+
+  let create ~file ~start ~end_ = { file; end_; start }
+
+  let of_location (location : Location.t) =
+    let loc_start = location.loc_start in
+    let loc_end = location.loc_end in
+    { file = loc_start.pos_fname
+      ; start = Location_in_file.create ~line:loc_start.pos_lnum
+          ~char:loc_start.pos_cnum ~bol:loc_start.pos_bol
+     ; end_ = Location_in_file.create ~line:loc_end.pos_lnum
+          ~char:loc_end.pos_cnum ~bol:loc_end.pos_bol
+    }
+
+  let file { file; _ } = file
+  let start { start; _ } = start
+  let end_ { end_; _ } = end_
+end
+
+type Format.stag += Location_mapping_tag of Range.t * string option
+
+let with_location_mapping ?label ~loc ppf f =
+  if is_tracking () then begin
+    let range = Range.of_location loc in
+    Format.pp_open_stag ppf (Location_mapping_tag (range, label));
+    let ret = f () in
+    Format.pp_close_stag ppf ();
+    ret
+  end else f ()
+
+module Mappings = struct
+  module Item = struct
+    type t = {
+      source : Range.t;
+      ir : Range.t;
+      label : string option;
+    }
+
+    let create ~source ~ir ~label = { source; ir; label }
+    let source { source; _ } = source
+    let ir { ir; _ } = ir
+    let label { label; _ } = label
+  end
+  type t = Item.t list
+
+  let empty = []
+
+  let append mapping t = mapping :: t
+
+  let print ppf t =
+    List.iter (fun Item.{source; ir; label} ->
+        Format.fprintf ppf "%s: %a -> %a\n"
```

Comment:
(Don't we want the colon to be present iff `label`
matches `Some _`?)

Reply:
Right, I'll fix this